### PR TITLE
Add signature v4

### PR DESCRIPTION
**Pull request by almp1**
Added signature of version 4 as per link below

http://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html

### CODE REVIEW

File: src/aws_credentials.pas [hunk 1]
```diff
@@ -19,13 +19,46 @@ interface
...
+  { TAWSSignatureHMAC256 }
+
+  TAWSSignatureHMAC256 = class sealed(TInterfacedObject, IAWSSignatureHMAC256)
+  private
+    FAccessKey: string;
+    FDataStamp: string;
+    FRegionName: string;
+    FServiceName: string;
+  public
+    constructor Create(const sAccessKey, sDataStamp, sRegionName, sServiceName: string);
```
Comment (mdbs99) [inline src/aws_credentials.pas]: @almp1 please remove these prefixes "s". Here and in all the code that you did the same.

File: src/aws_credentials.pas [hunk 2]
```diff
@@ -19,13 +19,46 @@ interface
...
+
 type
+
+  IAWSSignatureHMAC256 = interface(IInterface)
+  ['{9158D9A2-7ABA-4126-9F63-264E947AC60A}']
+    function AccessKey: string;
+    function DataStamp: string;
+    function RegionName: string;
+    function ServiceName: string;
+    function Signature: TSHA256Digest;
```
Comment (mdbs99) [inline src/aws_credentials.pas]: @almp1 why do we need this new type `TSHA256Digets`? Did you search if Synapse there is some similar?

File: src/aws_credentials.pas [hunk 3]
```diff
@@ -75,8 +108,164 @@   TAWSSignatureVersion3 = class sealed(TAWSAbstractSignature)
...
+  for i := 1 to sList.Count - 1 do
+    begin
+      sList.GetNameValue(i, sName, sValue);
+      sToSing := sToSing + LowerCase(sName) + ';';
+      sToCanonical := sToCanonical + LowerCase(sName)+':'+sValue+#10;
+    end;
+  system.Delete(sToSing, Length(sToSing), 1);
+end;
+
+function TAWSSignatureVersion4.Calculate(Request: IHTTPRequest): string;
```
Comment (mdbs99) [inline src/aws_credentials.pas]: @almp1 this method is complex. Is it possible to refactor, creating some other private methods?

**Review by mdbs99 — COMMENTED** [review]
I think all these 3rd files are open source and we can use them even in private projects, right? Make sure about it, please.

**Comment by mdbs99** [comment]
@almp1 when you to make these changes, please, send a new PR to `develop` branch. I will merge these new features to `master` in the future using `develop`. Thank you!

**Review by almp1 — COMMENTED** [review]
Yes the source code copied or changed under the Mozilla license must continue under this license. However, this code can be combined into a program with proprietary files. In addition, it is possible to create a proprietary version of a code under the Mozilla license

File: src/aws_credentials.pas (hunk 2 above)
Comment (almp1) [inline src/aws_credentials.pas]: The signature combination with hmac_sha256, was only found in synopse files

**Comment by mdbs99** [comment]
@almp1 Ok, thank you. I will merge your contributions as soon as possible as we have been talking on private.

**Review by mdbs99 — CHANGES_REQUESTED** [review]
@almp1 please do one more change: rename the directory "/terceiros" to "/3rd/synopse"

**Review by mdbs99 — APPROVED** [review]

**Comment by mdbs99** [comment]
@almp1 thanks. I merged into `develop` branch.

**Pull request closed without merging**